Detail view

Allcharles opened the pull request:
# Detail View Component

Redesigned `QuestionAnswerComponent` from PR #174 with improvements based on comments. Related issue: https://github.com/QutEcoacoustics/workbench-client/issues/193

## Changes

- Created `DetailViewComponent`
- Created `ViewComponent`
- Added humanize-duration library dependency (used to be length of audio files until luxon adds official support)
  - There is a typescript version of this library (https://github.com/Nightapes/HumanizeDuration.ts), however i did not add it as:
    - a) library is a fillin for missing luxon functionality which will hopefully come and
    - b) typescript version may go out of sync with original library as its maintained by a different user

## Issues

- Doesn't support association loading
- No `href`/`routerLink` support
- `TheirEditComponent` producing the following error when file input is added to form:

```typescript
ERROR DOMException: "An attempt was made to use an object that is not, or is no longer, usable"
```

## Visual Changes

Example `DetailViewComponent` (everything after `<hr />`)
![image](https://user-images.githubusercontent.com/3955116/78856994-87ffd080-7a6b-11ea-80b3-720760b2a017.png)

Allcharles committed:
Created components

Allcharles committed:
Initial look at DetailViewComponent

Allcharles committed:
Updated profile.json

Allcharles committed:
Added checkbox type

Allcharles committed:
Cleanup

atruskie commented on src/app/component/profile/profile.json:
```diff
@@ -0,0 +1,214 @@
+{
```
is this json schema?
is it valid?

it should probable have the `$schema` key in it, not least for utilization of vscode intellisense http://json-schema.org/schema#

Allcharles replied:
Currently I have not created any json schemas. Ill set it up in a different PR as I'm unsure how much will change to get it functional. Issue here: https://github.com/QutEcoacoustics/workbench-client/issues/181

atruskie commented on src/app/component/shared/checkbox/checkbox.component.spec.ts:
```diff
@@ -21,6 +21,9 @@ describe("CheckboxComponent", () => {
     expect(component).toBeTruthy();
   });
 
+  xit("should be centered", () => {});
```
probably worth a comment on why these are disabled

atruskie commented on src/app/component/shared/checkbox/checkbox.component.ts:
```diff
@@ -21,12 +21,13 @@ import {
   @Input() id: string;
   @Input() checked: boolean;
   @Input() disabled: boolean;
+  @Input() ngClass: { [key: string]: boolean } = { "mx-auto": true };
```
I'd prefer we didn't let class names leak into components. Can we do the fallback in the template rather than as a default  parameter here?

atruskie commented on src/app/component/shared/detail-view/detail-view.component.spec.ts:
```diff
@@ -0,0 +1,23 @@
+
+  it("should create", () => {
+    expect(component).toBeTruthy();
+  });
```
more tests are coming I presume?

atruskie commented on src/app/component/shared/detail-view/detail-view.component.ts:
```diff
@@ -0,0 +1,39 @@
+  selector: "app-detail-view",
+  template: `
+    <div>
+      <div *ngFor="let field of fields">
```
semantically, I'd say this is more reminiscent of a definition list https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dl

atruskie commented on src/app/component/shared/detail-view/detail-view.component.ts:
```diff
@@ -0,0 +1,39 @@
+            </p>
+          </div>
+          <div class="col-sm-9">
+            <app-view [view]="model[field.key]"></app-view>
```
`app-view` seems like a very general name. Are we viewing an entire application?

atruskie commented on src/app/component/shared/detail-view/view/view.component.ts:
```diff
@@ -0,0 +1,107 @@
+  selector: "app-view",
+  template: `
+    <ng-container *ngIf="!children; else hasChildren">
+      <pre *ngIf="styling === CodeStyling.Code">{{ display }}</pre>
```
`CodeStyling` perhaps is better named `FieldStyling`

atruskie commented on src/app/component/shared/detail-view/view/view.component.ts:
```diff
@@ -0,0 +1,107 @@
+
+    try {
+      this.styling = CodeStyling.Code;
+      this.display = JSON.stringify(value);
```
I'd like to use a tree view here to expose an arbitrary object graph.  You can even reuse this `view` component for the nodes!

This can be done later on.

atruskie replied:
Oh and that doesn't negate the utility of `CodeStyling.Code` - there are some code/script block we'll want to render as 'code'

Allcharles replied:
https://github.com/QutEcoacoustics/workbench-client/issues/183

atruskie commented on src/app/component/shared/detail-view/view/view.component.ts:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  /**
+   * Indicate answer is still loading
```
not an answer

atruskie commented on src/app/component/shared/detail-view/view/view.component.ts:
```diff
@@ -0,0 +1,107 @@
+import { toRelative } from "src/app/interfaces/apiInterfaces";
+
+@Component({
+  selector: "app-view",
```
```suggestion
  selector: "render-field",
```

Allcharles replied:
Needs to have the `app` at the start to remove chance of collision with libraries. But `app-render-view` works.

atruskie replied:
As per https://angular.io/guide/styleguide#component-custom-prefix yes it should be prefixed

but we should namespace it with our app identifier (not `app` that's too generic).

atruskie commented on src/app/interfaces/apiInterfaces.ts:
```diff
@@ -41,6 +42,35 @@ export type DateTimeTimezone = DateTime;
+  */
+  return seconds
+    ? Duration.fromObject({
+        years: 0,
+        months: 0,
+        days: 0,
+        hours: 0,
+        minutes: 0,
```
```suggestion
```

Allcharles replied:
https://github.com/moment/luxon/issues/687. This is an intended feature, I must have those keys there.

atruskie replied:
This is straight up counter intuitive. Should leave a comment behind at least (e.g. link to issue).

atruskie reviewed (changes requested):
Looks great. Unit tests still need work.

Allcharles committed:
Cleanup checkbox review comments

- https://github.com/QutEcoacoustics/workbench-client/pull/177#discussion_r405960461
- https://github.com/QutEcoacoustics/workbench-client/pull/177#discussion_r405960029

Allcharles committed:
Minor fixes to app render view

- https://github.com/QutEcoacoustics/workbench-client/pull/177#discussion_r405967180
- https://github.com/QutEcoacoustics/workbench-client/pull/177#discussion_r405967393
- https://github.com/QutEcoacoustics/workbench-client/pull/177#discussion_r405961208
- https://github.com/QutEcoacoustics/workbench-client/pull/177#discussion_r405962038

Allcharles committed:
Converted DetailView to definition list

https://github.com/QutEcoacoustics/workbench-client/pull/177#discussion_r405960942

Allcharles committed:
Added view component tests

Allcharles committed:
Created DetailViewComponent tests

Allcharles committed:
Fixed checkbox tests

Allcharles committed:
Merge branch 'master' into detail-view

Allcharles committed:
Fix imports

atruskie commented on src/app/component/shared/detail-view/detail-view.component.spec.ts:
```diff
@@ -1,23 +1,214 @@
+      it("should create answer", () => {
+        const answers = getAnswers();
+        expect(answers[0].innerText.trim()).toBe("0");
+      });
```
bad terminology here.

it should be `field` and `value` not `question` and `answer`

atruskie commented on src/app/component/shared/detail-view/detail-view.component.ts:
```diff
@@ -10,21 +10,14 @@ import { AbstractModel } from "src/app/models/AbstractModel";
+      <dt class="col-sm-3 text-left text-sm-right font-weight-bold">
+        {{ field.templateOptions.label }}
+      </dt>
+      <app-render-view
```
Rename to `baw-render-field` - we are not rendering a view

atruskie commented on src/app/component/shared/detail-view/detail-view.component.ts:
```diff
@@ -10,21 +10,14 @@ import { AbstractModel } from "src/app/models/AbstractModel";
+      <app-render-view
+        class="col-sm-9"
+        [view]="model[field.key]"
+      ></app-render-view>
```
```suggestion
      />
```

atruskie commented on src/app/component/shared/detail-view/view/view.component.spec.ts:
```diff
@@ -1,23 +1,325 @@
-describe("ViewComponent", () => {
-  let component: ViewComponent;
-  let fixture: ComponentFixture<ViewComponent>;
+describe("RenderViewComponent", () => {
```
rename to `RenderFieldComponent`

atruskie commented on src/app/component/shared/detail-view/view/view.component.ts:
```diff
@@ -3,29 +3,44 @@ import { DateTime, Duration } from "luxon";
 
 @Component({
-  selector: "app-view",
+  selector: "app-render-view",
```
```suggestion
  selector: "baw-render-field",
```

We are not doing anything with a "view" - we're just displaying the value of a field

Allcharles committed:
Address review comments

Allcharles committed:
Fix failing build
